Minimal native support

bsansouci opened the pull request:
Hey @bobzhang ! 
I realized that it'd probably better to make a PR with a minimal support of native compilation, rather than go down the rabbit hole of making Belt work, mostly because a minimal native support is already fun and useful.
It might look scary, but most changes are hidden under `#if BS_NATIVE then`. The only changes that aren't hidden under that flag are done to help make the diff more readable, and don't cause a semantic change. For example I pass the `backend` arg to all functions, which defaults to `Js`, but it's not used by any function when compiled with `BS_NATIVE=false`.
For native compilation, I have to nest the build artifacts under `lib/bs/js` / `lib/bs/bytecode` / `lib/bs/native` to make sure the files didn't step on each other (like the `cmi` which aren't the same between JS and native/bytecode). That nesting is also compiled away for bsb, of course, but I do pass `nested` (the directory name) to the functions that need them.

Here is a list of things that PR does. 
- Add functions to access the ocaml stdlib vendored in bucklescript (`Bsb_build_util`)
- Add some custom exceptions for native compilation.
- Add -bs-binary-simple-ast flag to bsc, which generates a simple AST file (which `ocamlc` / `ocamlopt` understands)
- Fix the `.merlin` file generation
- Adds `BS_NATIVE_PPX` to differentiate between the part of the ppx used
  inside bsc and the standalone ppx used by bsb-native for compiling
  belt to native.
- Make `Bsb_ninja_check` also check if the `-backend` argument changed to know to rebuild or not
- Fixed old merged BS_NATIVE stuff (including bsb_helper)
- Pass `backend` to all the functions, but compile out the `-backend` flag
- Add a whole new file called `bsb_ninja_native` which handles outputting the ninja file for building, linking and packing. Super similar to `Bsb_ninja_file_groups`.
- Modified `Bsb_ninja_gen` to add the global variables we need for native compilation and call the new `Bsb_ninja_native.handle_file_groups` function.
- Make `Bsb_world` collect a list of the dependencies in topological order (since we're already iterating over them).
- Simplify bsb_helper_depfile_gen
- Remove -bs-super-errors from linker/packer for now (that requires changing the compiler to accept that flag and enable support error)
- Make `Ext_bytes.escape` also escape `$`, I think because we might pass arguments that contain those characters.
- Update `scripts/ninja.js` to take a `-native` flag to generate the ninja files with `BS_NATIVE=true`. This allows me to build the whole thing running `ninja.js config -native && ninja.js build`. Making a release will require a couple more changes but those can be done in a subsequent PR.

bobzhang commented:
hi @bassjacob thanks for keeping working on this, I will do the review and get back to you asap

bobzhang commented on jscomp/bsb/bsb_build_util.ml:
```diff
@@ -244,3 +244,13 @@ let rec walk_all_deps_aux
+let get_ocaml_dir ~cwd =
+  (* TODO: fix this hardcoded version number *)
+  (Filename.dirname (get_bsc_dir ~cwd)) // "native" // "4.02.3"
+
```
this magic number can be replaced with Sys.ocaml_version with some processing

bobzhang commented on jscomp/bsb/bsb_config_parse.ml:
```diff
@@ -79,6 +79,11 @@ let extract_main_entries (map :json_map) =
+  match json with
+    | Obj {map} -> extract_main_entries map
+    | _ -> assert false
 
```
The json file is parsed several time, shall we add a lazy field to avoid parsing it several times like
```
let json_lazy = lazy (Ext_json_parse.parse_json_... )
```

bobzhang commented on jscomp/bsb/bsb_merlin_gen.ml:
```diff
@@ -86,18 +98,27 @@ let output_merlin_namespace buffer ns=
 
+#if BS_NATIVE then
+let bsc_flg_to_merlin_ocamlc_flg bsc_flags backend  =
+#else
```
could `backend` be made a lazy value so that the patch does not need be intrusive (the argument does not need be passed across each function), so is `nest` argument

```
let backend = lazy (...)
```

bobzhang commented on jscomp/bsb/bsb_merlin_gen.ml:
```diff
@@ -152,12 +173,30 @@ let merlin_file_gen ~cwd
+      | Bsb_config_types.Native   -> Literals.native
+      | Bsb_config_types.Bytecode -> Literals.bytecode
+    in
+#end
```
This is the same thing, we can push it into lazy evaluation since it only needs to be evaluated once during the lifetime of such process

bsansouci replied:
Now that it doesn't take any args and it still has a side effect, do you think I should use lazy? It might make sense, I'm just scared of having the issue of calling this too early, because the backend is parsed and be stuck with that wrong value.

bobzhang commented on jscomp/bsb/bsb_ninja_check.ml:
```diff
@@ -28,6 +28,7 @@ type t =
     source_directory :  string ;    
+    cmdline_backend: Bsb_config_types.compilation_kind_t;
   }
 
```
maybe we don't need add a new field, we can just change the name of file, for js backend its name is xx, for native backend its name is yy so they don't conflict with each other

bobzhang replied:
I am thinking to reduce such conflicts even for ninja file name, they can pick different names so less conflict is introduced, then you can call `ninja -f xx.ninja`

bobzhang commented on jscomp/bsb/bsb_ninja_rule.ml:
```diff
@@ -170,16 +188,19 @@ let make_custom_rules
+    if gen_simple then
+      Buffer.add_string buf " -bs-simple-binary-ast";
     Buffer.add_string buf " $bsc_flags -o $out -bs-syntax-only -bs-binary-ast $in";   
     Buffer.contents buf
```
how do you generate .d file in this case?

bobzhang commented on jscomp/bsb_helper/bsb_helper_packer.ml:
```diff
@@ -22,46 +22,85 @@
       )
     []
     sorted_tasks in
+  let warning_command = if String.length warnings > 0 then
```
note to help me review it easier, the changes to bsb_native alone could be sent separately and merged quickly

bobzhang commented on jscomp/ext/ext_bytes.ml:
```diff
@@ -55,6 +58,10 @@ let escaped s =
+      | '$' ->
+           Bytes.unsafe_set s' !n '$'; incr n; Bytes.unsafe_set s' !n '$'
+#end
       | '\n' ->
```
where is this function used in bs_native? The escaping rules does not seem correct to me

bobzhang replied:
If you read the docs of {String.ecaped}, it escape the special character following ocaml lexical convention which is not relevant to windows path convention, I guess you need a special function for windows path escaping

bobzhang commented on scripts/ninja.js:
```diff
@@ -1588,7 +1589,7 @@ function nativeNinja() {
-    command = $ocamlopt -safe-string -I +compiler-libs -opaque ${includes} -g -w +6-40-30-23 -warn-error +a-40-30-23 -absname -c $in
+    command = BS_NATIVE=${BS_NATIVE} $ocamlopt -safe-string -I +compiler-libs -opaque ${includes} -g -w +6-40-30-23 -warn-error +a-40-30-23 -absname -c $in
     description = $out : $in
 rule archive
```
where is this env variable read?

bsansouci commented:
Friendly ping @bobzhang  :)

bobzhang commented:
> Friendly ping @bobzhang :)

Looking into it

bobzhang commented on jscomp/bsb/bsb_ninja_check.ml:
```diff
@@ -63,16 +64,18 @@ let pp_check_result fmt (check_resoult : check_result) =
 
-let rec check_aux cwd (xs : string array) (ys: float array) i finish =
+let rec check_aux cwd ~nested (xs : string array) (ys: float array) i finish =
   if i = finish then Good
```
what's `nested`? can we make the diff as small as possible

bsansouci replied:
Good catch, that was very useless. I used your change to make this less intrusive (I think) by ~200 lines :D

bsansouci commented on jscomp/bsb/bsb_build_util.ml:
```diff
@@ -244,3 +244,20 @@ let rec walk_all_deps_aux
+let get_ocaml_lib_dir ~is_js ~cwd =
+  if is_js then (Filename.dirname (get_bsc_dir ~cwd)) // "lib" // "ocaml"
+  else (get_ocaml_dir ~cwd) // "lib" // "ocaml"
+#end
```
I changed this to be inside Bsb_global_paths! Thanks

bsansouci commented on jscomp/bsb_helper/bsb_helper_depfile_gen.ml:
```diff
@@ -271,64 +276,3 @@ let emit_d
-      | None -> 
-        raise (Arg.Bad ("don't know what to do with  " ^ fn))
-    end
-#end
```
This is part of my cleanup of old broken stuff.

bsansouci commented:
Updated the diff! It's quite a bit smaller thanks for that! 
I added Bsb_global_backend which encapsulates what we talked about. Tell me what you think.

bobzhang commented on jscomp/common/js_config.mli:
```diff
@@ -89,6 +89,9 @@ val sort_imports : bool ref
+#end
 
 val bs_suffix : bool ref
 val debug : bool ref
```
This does not have to be on conditional compilation, we can support it natively in a separate PR

bobzhang commented on jscomp/core/js_implementation.ml:
```diff
@@ -50,6 +50,13 @@ let process_with_gentype filename =
+     let oc = open_out_bin (outputprefix ^ Literals.suffix_mliast_simple) in 
+     Ml_binary.write_ast Mli !Location.input_name ast oc;
+     close_out oc ;
+   end;
```
same as above

bobzhang commented on jscomp/syntax/native_ast_derive_abstract.ml:
```diff
@@ -7,9 +7,16 @@ let deprecated name =
+#else
    if arg_name.[0] = '?' then
      String.sub arg_name 1 (String.length arg_name - 1)
    else arg_name
```
can you make it a separate PR

bsansouci commented:
Closing! Working on a simpler approach.
